File: src/onion/compiler/OnionClassLoader.java
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2005, Kota Mizushima, All rights reserved.       *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package onion.compiler;

import java.io.File;
import java.net.*;
import java.net.URL;
import java.net.URLClassLoader;

/**
 * @author Kota Mizushima
 * Date: 2005/07/19
 */
public class OnionClassLoader extends URLClassLoader {
  public OnionClassLoader(ClassLoader parent, String[] classPath, CompiledClass[] classes) 
    throws MalformedURLException {
    super(getURLs(classPath), parent);
    for(int i = 0; i < classes.length; i++){
      String className = classes[i].getClassName();
      byte[] content = classes[i].getContent();
      defineClass(className, content, 0, content.length);
    }
  }
  
  protected Class findClass(String name) throws ClassNotFoundException {
    return super.findClass(name);
  }
  
  private static URL[] getURLs(String[] classPath) throws MalformedURLException {
    URL[] urls = new URL[classPath.length];
    for(int i = 0; i < classPath.length; i++) {
      urls[i] = new File(classPath[i]).toURI().toURL();
    }
    return urls;
  }
}

File: src/main/scala/onion/compiler/OnionClassLoader.scala
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2016-, Kota Mizushima, All rights reserved.  *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package onion.compiler

import java.io.File
import java.net._
import java.net.URL
import java.net.URLClassLoader

/**
 * @author Kota Mizushima
 *
 */
class OnionClassLoader @throws(classOf[MalformedURLException]) (parent: ClassLoader, classPath: Seq[String], classes: Seq[CompiledClass]) extends
  URLClassLoader(classPath.map(cp => new File(cp).toURI.toURL).toArray, parent) {

  // Define classes lazily: eager defineClass in declaration order fails when
  // a subclass is declared before its superclass in the same script (the JVM
  // resolves the super while defining and it isn't there yet)
  private val pendingClasses = scala.collection.mutable.Map(classes.map(k => k.className -> k.content)*)

  classes.foreach(k => loadClass(k.className))

  @throws(classOf[ClassNotFoundException])
  protected override def findClass(name: String): Class[?] =
    pendingClasses.remove(name) match {
      case Some(content) => defineClass(name, content, 0, content.length)
      case None => super.findClass(name)
    }
}
